pkg/controllermanager/controller/shoot/shoot_reference_control.go

```diff
@@ -0,0 +1,213 @@
+	gardenlogger "github.com/gardener/gardener/pkg/logger"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func (c *Controller) shootReferenceAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.shootReferenceQueue.Add(key)
+}
+
+func (c *Controller) shootReferenceUpdate(oldObj, newObj interface{}) {
+	var (
+		oldShoot = oldObj.(*gardencorev1beta1.Shoot)
+		newShoot = newObj.(*gardencorev1beta1.Shoot)
+	)
+
+	if !apiequality.Semantic.Equalities.DeepEqual(oldShoot.Spec, newShoot.Spec) || newShoot.DeletionTimestamp != nil {
+		key, err := cache.MetaNamespaceKeyFunc(newObj)
+		if err != nil {
+			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", newObj, err)
+			return
+		}
+		c.shootReferenceQueue.Add(key)
+	}
+}
+
+// FinalizerName is the name of the finalizer used for the reference protection.
+const FinalizerName = "gardener.cloud/reference-protection"
+
+// NewShootReferenceReconciler creates a new instance of a reconciler which checks object references from shoot objects.
+func NewShootReferenceReconciler(clientMap clientmap.ClientMap, logger logrus.FieldLogger) reconcile.Reconciler {
+	return &shootReferenceReconciler{
+		clientMap: clientMap,
+		logger:    logger,
+	}
+}
+
+type shootReferenceReconciler struct {
+	clientMap clientmap.ClientMap
+	logger    logrus.FieldLogger
+}
+
+// Reconcile checks the shoot in the given request for references to further objects in order to protect the from
+// deletions as long as they are still referenced.
+func (r *shootReferenceReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.TODO()
+
+	gardenClient, err := r.clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := gardenClient.Client().Get(ctx, request.NamespacedName, shoot); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return reconcile.Result{}, err
+		}
+		r.logger.Debugf("shoot %q is not found, not trying to reconcile", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	r.logger.Infof("[SHOOT REFERENCE CONTROL] %s", request)
+
+	if err := r.reconcileShootReferences(ctx, shoot); err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *shootReferenceReconciler) reconcileShootReferences(ctx context.Context, shoot *gardencorev1beta1.Shoot) error {
+	gardenClient, err := r.clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return fmt.Errorf("failed to get garden client: %w", err)
+	}
```

Comment:
Can you pass the client instead of `Get`ting it again?
```suggestion
func (r *shootReferenceReconciler) reconcileShootReferences(ctx context.Context, c client.Client, shoot *gardencorev1beta1.Shoot) error {
```